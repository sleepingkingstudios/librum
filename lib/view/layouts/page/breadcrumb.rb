# frozen_string_literal: true

module View::Layouts
  # Component for rendering a page breadcrumb.
  class Page::Breadcrumb < ViewComponent::Base
    extend Forwardable

    # @param breadcrumb
    #   [View::Layouts::Page::Breadcrumbs::BreadcrumbConfiguration] the
    #   breadcrumb to render.
    def initialize(breadcrumb:)
      super()

      @breadcrumb = breadcrumb
    end

    def_delegators :@breadcrumb,
      :label,
      :url

    # @return [View::Layouts::Page::Breadcrumbs::BreadcrumbConfiguration] the
    #   breadcrumb to render.
    attr_reader :breadcrumb

    # @return [true, false] true if the breadcrumb is active, otherwise false.
    def active?
      breadcrumb.active
    end
  end
end
