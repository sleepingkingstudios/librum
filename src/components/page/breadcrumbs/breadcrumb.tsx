import * as React from 'react';
import {
  Link,
  useLocation,
} from 'react-router-dom';

import type { Breadcrumb } from './types';

const EmptyBreadcrumb = ({ label }: { label: string }): JSX.Element => (
  <li className="page-breadcrumb">
    <span className="page-breadcrumb-empty">{ label }</span>
  </li>
);

const BreadcrumbLink = (
  { label, url }: { label: string, url: string }
): JSX.Element => (
  <li className="page-breadcrumb">
    <Link to={url} className="page-breadcrumb-link">{ label }</Link>
  </li>
)

export const PageBreadcrumb = (
  { breadcrumb }: { breadcrumb: Breadcrumb }
): JSX.Element => {
  const { active, label } = breadcrumb;
  const { pathname } = useLocation();

  if (!('url' in breadcrumb) || active === false) {
    return (<EmptyBreadcrumb label={label} />);
  }

  const { url } = breadcrumb;

  if (active !== true && pathname === url) {
    return (<EmptyBreadcrumb label={label} />);
  }

  return (<BreadcrumbLink label={label} url={url} />);
};
