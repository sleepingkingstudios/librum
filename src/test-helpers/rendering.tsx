import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
