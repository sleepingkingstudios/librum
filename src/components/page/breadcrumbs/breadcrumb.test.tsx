import * as React from 'react';

import '@testing-library/jest-dom';

import { PageBreadcrumb } from './breadcrumb';
import type { Breadcrumb } from './types';
import { render } from '@test-helpers/rendering';

describe('<PageBreadcrumb />', () => {
  describe('with url: undefined', () => {
    const breadcrumb: Breadcrumb = { label: 'The Void' };

    it('should match the snapshot', () => {
      const { asFragment, getByText, queryByRole } = render(
        <PageBreadcrumb breadcrumb={breadcrumb} />,
        { router: true },
      );

      expect(getByText(breadcrumb.label)).toBeVisible();
      expect(queryByRole('link', { name: breadcrumb.label })).toBeNull();

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with url: value', () => {
    const breadcrumb: Breadcrumb = { label: 'The Void', url: '/the-void' };

    describe('when the current location does not match the url', () => {
      it('should match the snapshot', () => {
        const { asFragment, getByRole } = render(
          <PageBreadcrumb breadcrumb={breadcrumb} />,
          { router: true },
        );
        const link = getByRole('link', { name: breadcrumb.label });

        expect(link).toBeVisible();
        expect(link).toHaveAttribute('href', breadcrumb.url);

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the current location matches the url', () => {
      it('should match the snapshot', () => {
        const { asFragment, getByText, queryByRole } = render(
          <PageBreadcrumb breadcrumb={breadcrumb} />,
          {
            initialEntries: ['/the-void'],
            router: true,
          },
        );

        expect(getByText(breadcrumb.label)).toBeVisible();
        expect(queryByRole('link', { name: breadcrumb.label })).toBeNull();

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with url: value and active: false', () => {
    const breadcrumb: Breadcrumb = {
      active: false,
      label: 'The Void',
      url: '/the-void',
    };

    describe('when the current location does not match the url', () => {
      it('should match the snapshot', () => {
        const { asFragment, getByText, queryByRole } = render(
          <PageBreadcrumb breadcrumb={breadcrumb} />,
          { router: true },
        );

        expect(getByText(breadcrumb.label)).toBeVisible();
        expect(queryByRole('link', { name: breadcrumb.label })).toBeNull();

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the current location matches the url', () => {
      it('should match the snapshot', () => {
        const { asFragment, getByText, queryByRole } = render(
          <PageBreadcrumb breadcrumb={breadcrumb} />,
          {
            initialEntries: ['/the-void'],
            router: true,
          },
        );

        expect(getByText(breadcrumb.label)).toBeVisible();
        expect(queryByRole('link', { name: breadcrumb.label })).toBeNull();

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with url: value and active: true', () => {
    const breadcrumb: Breadcrumb = {
      active: true,
      label: 'The Void',
      url: '/the-void',
    };

    describe('when the current location does not match the url', () => {
      it('should match the snapshot', () => {
        const { asFragment, getByRole } = render(
          <PageBreadcrumb breadcrumb={breadcrumb} />,
          { router: true },
        );
        const link = getByRole('link', { name: breadcrumb.label });

        expect(link).toBeVisible();
        expect(link).toHaveAttribute('href', breadcrumb.url);

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the current location matches the url', () => {
      it('should match the snapshot', () => {
        const { asFragment, getByRole } = render(
          <PageBreadcrumb breadcrumb={breadcrumb} />,
          {
            initialEntries: ['/the-void'],
            router: true,
          },
        );
        const link = getByRole('link', { name: breadcrumb.label });

        expect(link).toBeVisible();
        expect(link).toHaveAttribute('href', breadcrumb.url);

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });
});
