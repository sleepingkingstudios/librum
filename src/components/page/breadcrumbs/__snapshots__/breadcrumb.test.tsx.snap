// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageBreadcrumb /> with url: undefined should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="page-breadcrumb"
  >
    <span
      class="page-breadcrumb-empty"
    >
      The Void
    </span>
  </li>
</DocumentFragment>
`;

exports[`<PageBreadcrumb /> with url: value and active: false when the current location does not match the url should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="page-breadcrumb"
  >
    <span
      class="page-breadcrumb-empty"
    >
      The Void
    </span>
  </li>
</DocumentFragment>
`;

exports[`<PageBreadcrumb /> with url: value and active: false when the current location matches the url should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="page-breadcrumb"
  >
    <span
      class="page-breadcrumb-empty"
    >
      The Void
    </span>
  </li>
</DocumentFragment>
`;

exports[`<PageBreadcrumb /> with url: value and active: true when the current location does not match the url should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="page-breadcrumb"
  >
    <a
      class="page-breadcrumb-link"
      href="/the-void"
    >
      The Void
    </a>
  </li>
</DocumentFragment>
`;

exports[`<PageBreadcrumb /> with url: value and active: true when the current location matches the url should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="page-breadcrumb"
  >
    <a
      class="page-breadcrumb-link"
      href="/the-void"
    >
      The Void
    </a>
  </li>
</DocumentFragment>
`;

exports[`<PageBreadcrumb /> with url: value when the current location does not match the url should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="page-breadcrumb"
  >
    <a
      class="page-breadcrumb-link"
      href="/the-void"
    >
      The Void
    </a>
  </li>
</DocumentFragment>
`;

exports[`<PageBreadcrumb /> with url: value when the current location matches the url should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="page-breadcrumb"
  >
    <span
      class="page-breadcrumb-empty"
    >
      The Void
    </span>
  </li>
</DocumentFragment>
`;
